### ISSUE TITLE
Linear eval for ResNet50?

### ISSUE THREAD

**Issue by droidnesis** [facebookresearch/dino#84]
The eval_linear.py doesnt seem to support ResNet50. How was it evaluated for the paper? 

Any pointers to changes needed to make it work for resnet50? 

Thanks for making this work open-source! <3

**Comment by RiSaMa** [comment]
Also interested in this question. 

I am assuming that:
- ResNet50, once trained, will output features with dimension 2048
- LinearClassifier will be: LinearClassifier(2048, num_labels=len(classes))

The below two lines should be used for loading the weights in the ResNet50:
```
# remove `module.` prefix
checkpoint = {k.replace("module.", ""): v for k, v in checkpoint.items()}
# remove `backbone.` prefix induced by multicrop wrapper
checkpoint = {
    k.replace("backbone.", ""): v for k, v in checkpoint.items()
}
```

Still some doubts:
- Should the DINOHead be discarded? So is it only used during training for training the ResNet50?


Does this reasoning make sense? Am I missing something important?

**Comment by 4knahs** [comment]
As far as I understood it the weights would be from the teacher and the head is discarded (at least this seems to be the case for ViT).

The linear eval code trains a linear classifier appended to some of the last transformer blocks which are in eval mode. For ResNet perhaps freezing the model weights and retraining the last FC layer is enough.

If one of the authors could confirm would be nice, along with the respective optimisations.

**Comment by 4knahs** [comment]
With the above approach and without major changes to the eval_linear pipeline I am only able to achieve 74.5% compared to the paper 75.3%. 

I am using this repo pretrained model for Resnet50, with `--nproc_per_node=8` and replicating the following args (note that some are coming from the train args and not used):

```
{
    "arch": "resnet50",
    "out_dim": 60000,
    "norm_last_layer": true,
    "warmup_teacher_temp": 0.04,
    "teacher_temp": 0.07,
    "warmup_teacher_temp_epochs": 50,
    "use_fp16": false,
    "weight_decay": 0.000001,
    "weight_decay_end": 0.000001,
    "clip_grad": 0,
    "batch_size_per_gpu": 64,
    "epochs": 100,
    "freeze_last_layer": 1,
    "lr": 0.3,
    "warmup_epochs": 10,
    "min_lr": 0.000001,
    "global_crops_scale": [0.14, 1.0],
    "local_crops_scale": [0.05, 0.14],
    "local_crops_number": 6,
    "seed": 0,
    "num_workers": 10,
    "world_size": 8,
    "ngpus": 8,
    "nodes": 1,
    "optimizer": "lars",
    "momentum_teacher": 0.996,
    "use_bn_in_head": true}
```

Spot anything wrong or are there any tricks i might be missing? Any particular seed i should be trying?

Thanks!

**Comment by mathildecaron31** [comment]
I use a learning rate of 0.3 and a total batch size of 256 (2 gpus, batch size per gpu is 128).
You can take a look at the eval logs for ResNet-50: https://dl.fbaipublicfiles.com/dino/dino_resnet50_pretrain/dino_resnet50_pretrain_eval_linear_log.txt

I can adapt the linear code to train with RN50. Alternatively you can also use the code from swav: https://github.com/facebookresearch/swav/blob/master/eval_linear.py

There is no particular trick :)

**Comment by mathildecaron31** [comment]
https://github.com/facebookresearch/dino/commit/9085367cf5186b29f69c85eb7fdacc3057094031

**Comment by 4knahs** [comment]
Thanks Mathilde for the quick reply and code. 

A fresh clone and that command results in 75.034% which is much closer.
